### PR TITLE
[SPARK-21513][SQL][FOLLOWUP] Allow UDF to_json support converting MapType to json for PySpark and SparkR

### PR DESCRIPTION
**Pull request by goldmedal**
## What changes were proposed in this pull request?
In previous work SPARK-21513, we has allowed `MapType` and `ArrayType` of `MapType`s convert to a json string but only for Scala API. In this follow-up PR, we will make SparkSQL support it for PySpark and SparkR, too. We also fix some little bugs and comments of the previous work in this follow-up PR.

### For PySpark
```
>>> data = [(1, {"name": "Alice"})]
>>> df = spark.createDataFrame(data, ("key", "value"))
>>> df.select(to_json(df.value).alias("json")).collect()
[Row(json=u'{"name":"Alice")']
>>> data = [(1, [{"name": "Alice"}, {"name": "Bob"}])]
>>> df = spark.createDataFrame(data, ("key", "value"))
>>> df.select(to_json(df.value).alias("json")).collect()
[Row(json=u'[{"name":"Alice"},{"name":"Bob"}]')]
```
### For SparkR
```
# Converts a map into a JSON object
df2 <- sql("SELECT map('name', 'Bob')) as people")
df2 <- mutate(df2, people_json = to_json(df2$people))
# Converts an array of maps into a JSON array
df2 <- sql("SELECT array(map('name', 'Bob'), map('name', 'Alice')) as people")
df2 <- mutate(df2, people_json = to_json(df2$people))
```
## How was this patch tested?
Add unit test cases.

cc @viirya @HyukjinKwon

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

File: R/pkg/R/functions.R [hunk 1]
```diff
@@ -1700,8 +1701,9 @@ setMethod("to_date",
           })
 
 #' @details
-#' \code{to_json}: Converts a column containing a \code{structType} or array of \code{structType}
-#' into a Column of JSON string. Resolving the Column can fail if an unsupported type is encountered.
+#' \code{to_json}: Converts a column containing a \code{structType}, array of \code{structType},
+#  a \code{mapType} or array of \code{mapType} into a Column of JSON string.
```
Comment (HyukjinKwon) [inline R/pkg/R/functions.R]: Looks `'` is missed at the first `#`.

File: python/pyspark/sql/functions.py [hunk 2]
```diff
@@ -1921,10 +1921,12 @@ def from_json(col, schema, options={}):
 @since(2.1)
 def to_json(col, options={}):
     """
-    Converts a column containing a [[StructType]] or [[ArrayType]] of [[StructType]]s into a
-    JSON string. Throws an exception, in the case of an unsupported type.
+    Converts a column containing a [[StructType]], [[ArrayType]] of [[StructType]]s,
+    a [[MapType]] or [[ArrayType]] of [[MapType]] into a JSON string.
+    Throws an exception, in the case of an unsupported type.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: While we are here, let's fix `[[StructType]]` to `` :class:`StructType` `` (and the same instances too) to make Python API documentation pretty.

File: R/pkg/R/functions.R [hunk 3]
```diff
@@ -1715,7 +1717,15 @@ setMethod("to_date",
...
-#' df2 <- mutate(df2, people_json = to_json(df2$people))}
+#' df2 <- mutate(df2, people_json = to_json(df2$people))
+#'
+#' # Converts a map into a JSON object
+#' df2 <- sql("SELECT map('name', 'Bob')) as people")
+#' df2 <- mutate(df2, people_json = to_json(df2$people))
+#'
+#' # Converts an array of maps into a JSON array
+#' df2 <- sql("SELECT array(map('name', 'Bob'), map('name', 'Alice')) as people")
+#' df2 <- mutate(df2, people_json = to_json(df2$people))
```
Comment (HyukjinKwon) [inline R/pkg/R/functions.R]: And.. missing closing parentheses for `\dontrun{`.

Comment (felixcheung) [reply R/pkg/R/functions.R]: ... meaning `}`

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out [hunk 4]
```diff
@@ -26,13 +26,13 @@ Extended Usage:
        {"time":"26/08/2015"}
       > SELECT to_json(array(named_struct('a', 1, 'b', 2));
        [{"a":1,"b":2}]
-      > SELECT to_json(map('a',named_struct('b',1)));
+      > SELECT to_json(map('a', named_struct('b', 1)));
```
Comment (viirya) [inline sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: I think you committed unrelated change?

Comment (viirya) [reply sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: Or you forget to commit `json-functions.sql`?

Comment (goldmedal) [reply sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: umm. I modified `ExpressionDescription` of `StructsToJson` at @HyukjinKwon 's suggestions which didn't be merged in last PR. Here's the test for `describe function extended to_json`, so I needed to regenerate the golden file for it. So this change isn't from `json-functions.sql`.

File: R/pkg/R/functions.R (hunk 3 above)
Comment (goldmedal) [reply R/pkg/R/functions.R]: ok  Thanks for careful review :)

File: sql/core/src/test/resources/sql-tests/results/json-functions.sql.out (hunk 4 above)
Comment (viirya) [reply sql/core/src/test/resources/sql-tests/results/json-functions.sql.out]: Oh. I see.

File: python/pyspark/sql/functions.py (hunk 2 above)
Comment (goldmedal) [reply python/pyspark/sql/functions.py]: ok Thanks.

File: python/pyspark/sql/functions.py [hunk 5]
```diff
@@ -1921,10 +1921,12 @@ def from_json(col, schema, options={}):
 @since(2.1)
 def to_json(col, options={}):
     """
-    Converts a column containing a [[StructType]] or [[ArrayType]] of [[StructType]]s into a
-    JSON string. Throws an exception, in the case of an unsupported type.
+    Converts a column containing a :class:`StructType, :class:`ArrayType` of :class:`StructType`s,
```
Comment (viirya) [inline python/pyspark/sql/functions.py]: Missing a ` after StructType.

**Comment by viirya** [comment]
LGTM except for one comment left.

**Comment by viirya** [comment]
retest this please.

File: python/pyspark/sql/functions.py [hunk 6]
```diff
@@ -1921,10 +1921,12 @@ def from_json(col, schema, options={}):
 @since(2.1)
 def to_json(col, options={}):
     """
-    Converts a column containing a [[StructType]] or [[ArrayType]] of [[StructType]]s into a
-    JSON string. Throws an exception, in the case of an unsupported type.
+    Converts a column containing a :class:`StructType`, :class:`ArrayType` of :class:`StructType`s,
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: I believe `` :class:`StructType`s `` should be ``:class:`StructType`\\s``. Currently, this is rendered as below:

File: python/pyspark/sql/functions.py [hunk 7]
```diff
@@ -1921,10 +1921,12 @@ def from_json(col, schema, options={}):
...
+    a :class:`MapType` or :class:`ArrayType` of :class:`MapType` into a JSON string.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/functions.py]: `` `:class:`MapType` `` -> `` :class:`MapType`\\s `` for consistency.

**Comment by goldmedal** [comment]
@HyukjinKwon @felixcheung @viirya 
I has finished those change at your suggestions for this PR and it also passed all tests. Please take a look when you are available. Thanks :)

**Review by HyukjinKwon — COMMENTED** [review]
LGTM @felixcheung, do you maybe have more comments?

File: python/pyspark/sql/functions.py (hunk 6 above)
Comment (goldmedal) [reply python/pyspark/sql/functions.py]: ok, I'll modified it.  Because I'm not really familiar with python, thanks for your suggestions. :)

Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py]: That's fine :).

**Comment by viirya** [comment]
LGTM

**Comment by felixcheung** [comment]
AppVeyor didn't run on this?

**Comment by HyukjinKwon** [comment]
D'oh, yes. I wonder why it was not triggered. I manually triggered via my account:

Build started: [SparkR] `ALL` [PR-19223](https://ci.appveyor.com/project/spark-test/spark/branch/8981A5F1-E2DC-4015-8266-12E9ADEE189B)
Diff: https://github.com/apache/spark/compare/master...spark-test:8981A5F1-E2DC-4015-8266-12E9ADEE189B

**Comment by goldmedal** [comment]
@HyukjinKwon Thanks for triggering AppVeyor. In normal case, will AppVeyor be triggered automatically?

**Comment by HyukjinKwon** [comment]
Yes, when there are some changes in:

https://github.com/apache/spark/blob/828fab03567ecc245a65c4d295a677ce0ba26c19/appveyor.yml#L29-L35

It should run the R tests on Windows via AppVeyor.

**Comment by goldmedal** [comment]
ok. I got it. Thanks :)

**Comment by HyukjinKwon** [comment]
Looks passed fine. Let me merge this one.

**Comment by viirya** [comment]
Thanks @felixcheung @HyukjinKwon

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by goldmedal** [comment]
Thanks @HyukjinKwon @felixcheung @viirya
